Create TopologicalSort page

MarioAndWario opened the pull request:
Add introduction to topological sorting.
Add BFS and DFS templates.
Add 3 examples: Course Schedules I/II, Alien Dictionary, Minimum Height Trees.
Add exercises.
Add references.

fuxuemingzhu committed:
Merge pull request #37 from ojeveryday/docsify

Docsify 新增自动目录和 GitTalk 功能

fuxuemingzhu committed:
Merge pull request #38 from ojeveryday/docsify

增加 Gitalk

fuxuemingzhu committed:
Merge pull request #39 from ojeveryday/docsify

增加 Gitalk

fuxuemingzhu committed:
Merge pull request #41 from ojeveryday/docsify

修改 docsify 中的 id 为文件路径

fuxuemingzhu committed:
Merge pull request #45 from ojeveryday/docsify

修改 docsfiy 中的 id 是 URI 的 decode

fuxuemingzhu committed:
Merge pull request #48 from ojeveryday/docsify

docsify 增加 id 和 title

fuxuemingzhu committed:
Merge pull request #51 from ojeveryday/docsify

测试 id 和 title

fuxuemingzhu committed:
Merge pull request #52 from ojeveryday/docsify

恢复 gitalk 的 title 和 id

fuxuemingzhu committed:
Merge pull request #55 from ojeveryday/docsify

使用 md5作为 gitalk

fuxuemingzhu committed:
Merge pull request #57 from ojeveryday/docsify

增加 title

fuxuemingzhu committed:
Merge pull request #58 from ojeveryday/docsify

id 和 titile 修改

fuxuemingzhu committed:
Merge pull request #59 from ojeveryday/docsify

使用 md5

fuxuemingzhu committed:
Merge pull request #64 from ojeveryday/docsify

精简 gitalk 的 js

fuxuemingzhu committed:
Merge pull request #65 from ojeveryday/docsify

精简 gitalk 的 js

fuxuemingzhu committed:
Merge pull request #66 from ojeveryday/docsify

精简文件

fuxuemingzhu committed:
Merge pull request #69 from ojeveryday/docsify

gitalk 使用相对路径

fuxuemingzhu committed:
Merge pull request #73 from ojeveryday/docsify

hash reload

fuxuemingzhu committed:
认领 TopologicalSort

MarioAndWario committed:
Add BFS Python template

MarioAndWario committed:
1. Add BFS and DFS results as plots.
2. Add DFS python3 template.

MarioAndWario committed:
Add a simple application of topological sort.

MarioAndWario committed:
Finish introduction and code templates. Start working on problems.

MarioAndWario committed:
Merge branch 'ojeveryday:master' into TopologicalSort_wu2meng3

MarioAndWario committed:
Add description of Kahn's algorithm.
Add example of Alien Dictionary (LintCode 892)

MarioAndWario committed:
1. Add solutions of Course Schedule and Course Schedule II.
2. Add exercises.
3. Add references.

MarioAndWario committed:
1. Reorganize figures, add snapshots of BFS and DFS workflow.

MarioAndWario committed:
Cosmetic change

MarioAndWario committed:
Cosmetic change. Correct many typos.

MarioAndWario committed:
Cosmetic change

MarioAndWario committed:
Merge branch 'TopologicalSort_wu2meng3' of https://github.com/wu2meng3/AlgoWiki into TopologicalSort_wu2meng3

MarioAndWario committed:
cosmetic change

MarioAndWario committed:
Add Java and C++ solution for LC 210.

MarioAndWario committed:
Use default path for figures.

MarioAndWario committed:
Add Java and C++ solutions for LintCode 892 Alien Dictionary.

MarioAndWario committed:
Add Java and Python DFS solutions to LeetCode 210 Course Schedule II.

MarioAndWario committed:
Added LeetCode 310 as the 3nd example of topological sort. Modified the exercise list.

liweiwei1419 commented on TopologicalSort/README.md:
```diff
@@ -0,0 +1,1119 @@
+
+作者：wu2meng3
+
+拓扑排序是一个有向无环图（ directed acyclic graph ，简称 DAG ）所有结点的线性序列，满足以下两个条件：
```
有向无环图，首字母需要大写：Directed Acyclic Graph。

MarioAndWario replied:
好的，已经大写。

liweiwei1419 commented on TopologicalSort/README.md:
```diff
@@ -0,0 +1,1119 @@
+
+作者：wu2meng3
+
+拓扑排序是一个有向无环图（ directed acyclic graph ，简称 DAG ）所有结点的线性序列，满足以下两个条件：
```
定义需要使用一些经典教程中的描述，这里说：「拓扑排序」是一个「线性序列」是不好的。我查了一下百度百科（虽然也不一定权威），但是它的描述说得通：

> 对一个有向无环图（Directed Acyclic Graph 简称 DAG）G 进行拓扑排序，是将 G 中所有顶点排成一个线性序列，使得……

MarioAndWario replied:
已经按照百度百科修改了拓扑排序的定义。

liweiwei1419 commented on TopologicalSort/README.md:
```diff
@@ -0,0 +1,1119 @@
+
+其拓扑排序可以是: `0->1->2->3->4->5`，也可以是 `0->3->2->5->1->4`。
+
+![TopoSort](TopoSort.png)
```
下面的两个看起来线性的图是什么意思，需要和读者解释一下。

MarioAndWario replied:
已经将原来的图一分成了两个图，分别做了说明。

liweiwei1419 commented on TopologicalSort/README.md:
```diff
@@ -0,0 +1,1119 @@
+
+#### 思路
+
+该算法也叫 [Kahn 算法](https://en.wikipedia.org/wiki/Topological_sorting#Kahn's_algorithm)。首先，我们需要介绍**入度**（ indegree ）的概念。在有向图当中，一个结点的入度表示有多少条边指向它。对应的还有一个概念叫**出度** （ outdegree ），代表从一个结点出发有多少条边，但是在拓扑排序当中我们只会用到入度。入度为 0 的点的集合将作为 BFS 的起点。如果没有入度为 0 的点，则无法进行拓扑排序。BFS 由先进先出 （First In, First Out, FIFO）的队列 （ queue ）数据结构来实现。BFS 算法实现拓扑排序的伪代码如下：
```
宽度优先遍历的做法其实是很直观的，建议先把 BFS 的解法先进行直观描述，例如：先拿出不被任何节点指向的节点（入度为 0 ）的节点，然后把它从图中删去（从它出发、指向的所有节点的入度 -1）。把后面「为了形象的理解 BFS 的过程……」这部分提前。先讲解想法和思路，然后再给出代码和模板会好一些。

MarioAndWario replied:
同意！已经将思路和流程图放到代码之前。

liweiwei1419 commented on TopologicalSort/README.md:
```diff
@@ -0,0 +1,1119 @@
+
+#### 思路
+
+一句话总结就是**拓扑排序就是图后序遍历的倒序输出**。我们现在将这句话展开来理解。根据《算法导论》第 22 章，我们可以在 DFS 的过程中给每个结点盖两个时间戳，分别是发现这个结点的时间 `discoverTime[u]` 和遍历完该结点所有邻接结点的时间 `finishTime[u]` ，其中 `u` 代表一个结点。对于结点 `u`，DFS 算法遍历完其所有子结点之后才会判断其遍历结束，更新时间戳 `finishTime[u]` 。可以证明如果图中存在一条从 `u` 指向 `v` 的边，那么 `finishTime[v] < finishTime[u]` 。所以，按照 `finishTime` 的逆序输出结点就可以得到拓扑排序之后的序列。该算法有点类似回溯算法，我们需要在递归之前和递归之后更新每个结点的状态。每个结点有三种的可能的状态，分别是：
```
「拓扑排序就是图后序遍历的倒序输出」这个结论我只在 liuyubobobo 老师的课上看到过，对于初学的读者来说，看图理解文字是比较费脑的，前面可以加上一些说明，例如：对于拓扑排序而言，读者需要熟练掌握 BFS 解法（更常见、且相对思路自然），而 DFS 解法作为补充，了解即可。

MarioAndWario replied:
同意！已在DFS算法标题下做了说明。

MarioAndWario committed:
1. Capitalized `Directed Acyclic Graph`.
2. Used Baidu Baike's definition of topological sorting.
3. Splited the first plot into two such that the linear series are more understandable.
4. Added more description of BFS algorithm before presenting the code template.
5. Added discussion that DFS is not required in most cases.

MarioAndWario committed:
`queue` is for all nodes with indegree = 0.

MarioAndWario committed:
Cosmetic change

1. Add space before and after `BFS`.
2. queue should be `queue`.

MarioAndWario committed:
Shrinked the first plot.

MarioAndWario committed:
Reorganize the pseudocode of BFS

MarioAndWario committed:
Cosmetic change

MarioAndWario committed:
Cosmetic change
